prov-api/helpers/prov_connector.py

```diff
@@ -0,0 +1,1597 @@
+            for rel in path.relationships:
+                # pull nodes
+                start: graph.Node = rel.start_node
+                end: graph.Node = rel.end_node
+
+                # get ids
+                start_id = start.get(IDENTIFIER_TAG)
+                end_id = end.get(IDENTIFIER_TAG)
+
+                # check start/end in graph
+                if not digraph.has_node(start_id):
+                    # add start node into digraph
+                    digraph.add_node(start_id, **{
+                        'item_category': start.get('item_category'),
+                        'item_subtype': start.get('item_subtype'),
+                    })
+                if not digraph.has_node(end_id):
+                    # add end node into digraph
+                    digraph.add_node(end_id, **{
+                        'item_category': end.get('item_category'),
+                        'item_subtype': end.get('item_subtype'),
+                    })
+                if not digraph.has_edge(start_id, end_id):
+                    digraph.add_edge(start_id, end_id, **{
+                        'type': rel.type
+                    })
+
+    return digraph
+
+
+def special_contributing_dataset_query(starting_id: str, depth: int, config: Config) -> Dict[str, Any]:
+    """
+    Executes a special query to find contributing datasets up to a specified depth.
+
+    This function queries the Neo4j database for datasets that contribute to the
+    specified starting node, up to the given depth in the graph.
+
+    Args:
+        starting_id (str): The ID of the starting node.
+        depth (int): The maximum depth to traverse in the graph.
+        config (Config): Configuration object for database connection.
+
+    Returns:
+        Dict[str, Any]: A dictionary representing the graph in a D3.js-compatible format.
+
+    Raises:
+        RuntimeError: If attempting to query a mocked database.
+        HTTPException: If there's an error executing the query.
+    """
+    if config.mock_graph_db:
+        raise RuntimeError(
+            "Asking for real data from graph DB during mock! Returning [].")
+
+    # The cypher query to make for upstream lineage - only concerned with datasets upstream
+    query = f"""
+    MATCH r=((parent : Entity {{`item_subtype`:'DATASET'}}) <-[*1..{depth}]-(start{{`{IDENTIFIER_TAG}`:'{starting_id}'}})) RETURN r
+    """
+
+    # Run the query
+    try:
+        result = run_query(query, config)
+    except Exception as e:
+        raise HTTPException(
+            status_code=500,
+            detail=f"Neo4j lineage query experienced an exception: {e}."
+        )
+
+    # Convert to list of paths by pulling out the 'r' field from the RETURN r in above
+    # cypher query
+    paths: List[Optional[graph.Path]] = [record.get('r') for record in result]
+
+    # Apply simple exploration algorithm to traverse paths and produce networkx digraph
+    graph = generate_networkx_graph_from_lineage_paths(
```

Comment:
can you please rename the graph variable here on line 289 to something slightly different so my linter doesnt confuse it with the module imported from neo4j in line 34, otherwise nice work

Reply:
and same in other function(s) below